cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnelService.java

```diff
@@ -156,95 +205,81 @@ private void startTunnel(final String tunnelId, final JSONObject config, boolean
     try {
-      vpnTunnel.connectTunnel(config.getString("host"), config.getInt("port"),
-          config.getString("password"), config.getString("method"), remoteUdpForwardingEnabled);
+      vpnTunnel.connectTunnel(proxyConfig.host, proxyConfig.port, proxyConfig.password,
+          proxyConfig.method, remoteUdpForwardingEnabled);
     } catch (Exception e) {
       LOG.log(Level.SEVERE, "Failed to connect the tunnel", e);
-      onVpnStartFailure(OutlinePlugin.ErrorCode.VPN_START_FAILURE);
-      return;
+      tearDownActiveTunnel();
+      return OutlinePlugin.ErrorCode.VPN_START_FAILURE;
     }
-
-    broadcastVpnStart(OutlinePlugin.ErrorCode.NO_ERROR);
     startForegroundWithNotification(config);
-    storeActiveTunnel(tunnelId, config, remoteUdpForwardingEnabled);
+    storeActiveTunnel(config, remoteUdpForwardingEnabled);
+    return OutlinePlugin.ErrorCode.NO_ERROR;
   }
 
-  /**
-   * Tears down a tunnel started by calling |startTunnel|. Stops tun2socks, shadowsocks, and
-   * the system-wide VPN.
-   *
-   * @param tunnelId unique identifier for the tunnel.
-   * @throws IllegalArgumentException if |tunnelId| is missing.
-   * @throws IllegalStateException if the tunnel represented by |tunnelId| is not active.
-   */
-  public void stopTunnel(final String tunnelId) {
-    if (tunnelId == null) {
-      throw new IllegalArgumentException("Must provide a tunnel ID.");
-    } else if (!tunnelId.equals(activeTunnelId)) {
-      throw new IllegalStateException(
-          String.format(Locale.ROOT, "Tunnel %s not active.", tunnelId));
-    }
-    broadcastVpnStop();
+  private synchronized OutlinePlugin.ErrorCode stopTunnel(final String tunnelId) {
+    if (!isTunnelActive(tunnelId)) {
+      return OutlinePlugin.ErrorCode.UNEXPECTED;
+    }
     tearDownActiveTunnel();
+    return OutlinePlugin.ErrorCode.NO_ERROR;
   }
 
-  /**
-   * Determines whether a tunnel is active.
-   *
-   * @param tunnelId unique identifier for the tunnel.
-   * @throws IllegalArgumentException if |tunnelId| is missing.
-   * @return boolean indicating whether the tunnel is active.
-   */
-  public boolean isTunnelActive(final String tunnelId) {
-    if (tunnelId == null) {
-      throw new IllegalArgumentException("Must provide a tunnel ID.");
+  private synchronized boolean isTunnelActive(final String tunnelId) {
+    if (tunnelConfig == null || tunnelConfig.id == null) {
+      return false;
     }
-    return tunnelId.equals(activeTunnelId);
+    return tunnelConfig.id.equals(tunnelId);
   }
 
-  /* Helper method to broadcast a VPN start the failure and reset the service state. */
-  private void onVpnStartFailure(OutlinePlugin.ErrorCode errorCode) {
-    broadcastVpnStart(errorCode);
-    tearDownActiveTunnel();
+  private boolean isTunnelReachable(final String host, final int port) {
+    boolean isReachable = true;
+    try {
+      Shadowsocks.checkServerReachable(host, port);
+    } catch (Exception e) {
+      isReachable = false;
```

Comment:
No need for `isReachable`.

```suggestion
      return false;
```

Reply:
Done in 86eea94.